File: projectforge-business/src/test/java/org/projectforge/business/employee/EmployeeTest.java
package org.projectforge.business.employee;

import static org.testng.Assert.*;

import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;

import org.projectforge.business.fibu.EmployeeDO;
import org.projectforge.business.fibu.EmployeeDao;
import org.projectforge.business.fibu.Gender;
import org.projectforge.business.fibu.GenderConverter;
import org.projectforge.framework.persistence.history.DisplayHistoryEntry;
import org.projectforge.test.AbstractBase;
import org.projectforge.test.AbstractTestNGBase;
import org.springframework.beans.factory.annotation.Autowired;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

public class EmployeeTest extends AbstractTestNGBase
{

  @Autowired
  private EmployeeDao employeeDao;

  private List<EmployeeDO> employeeList;

  @BeforeClass
  public void init()
  {
    logon(AbstractBase.TEST_FULL_ACCESS_USER);
    employeeList = employeeDao.internalLoadAll();
    assertTrue(employeeList.size() > 0, "Keine Mitarbeiter in der Test DB!");
  }

  @AfterMethod
  public void clean()
  {
    //Get initial infos
    log.info("Cleanup deleted employess -> undelete");
    log.info("Count employees: " + employeeList.size());
    assertTrue(employeeList.size() > 0);
    for (EmployeeDO e : employeeList) {
      log.info("Employee: " + e.toString());
      if (e.isDeleted()) {
        //Undelete
        employeeDao.internalUndelete(e);
      }
    }
  }

  @Test
  public void testBirthday()
  {
    //Get initial infos
    log.info("Count employees: " + employeeList.size());
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    log.info("Employee: " + e.toString());

    List<DisplayHistoryEntry> historyEntries = employeeDao.getDisplayHistoryEntries(e);
    log.info("Employee history entry size: " + historyEntries.size());

    //Update employee
    Calendar birthday = new GregorianCalendar(1985, Calendar.DECEMBER, 17);
    e.setBirthday(birthday.getTime());
    employeeDao.update(e);

    //Check updates
    EmployeeDO updatdEmployee = employeeDao.getById(e.getId());
    assertEquals(updatdEmployee.getBirthday().getTime(), e.getBirthday().getTime());

    // test history
    List<DisplayHistoryEntry> updatedHistoryEntries = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(updatedHistoryEntries.size(), historyEntries.size() + 1);

    //Remove data and check
    e.setBirthday(null);
    employeeDao.update(e);

    //Check updates
    EmployeeDO updatdEmployeeRemove = employeeDao.getById(e.getId());
    assertNull(updatdEmployeeRemove.getBirthday());

    // test history
    List<DisplayHistoryEntry> updatedHistoryRemoveEntries = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(updatedHistoryRemoveEntries.size(), historyEntries.size() + 2);
  }

  @Test
  public void testMarkAsDeleted()
  {
    //Get initial infos
    log.info("Count employees: " + employeeList.size());
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    log.info("Employee: " + e.toString());

    //Mark as deleted
    employeeDao.markAsDeleted(e);

    //Check updates
    EmployeeDO updatdEmployee = employeeDao.getById(e.getId());
    assertTrue(updatdEmployee.isDeleted());

    employeeDao.update(e);
  }

  @Test
  public void testGender()
  {
    //Get initial infos
    log.info("Count employees: " + employeeList.size());
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    log.info("Employee: " + e.toString());
    List<DisplayHistoryEntry> historyEntriesBefore = employeeDao.getDisplayHistoryEntries(e);

    e.setGender(Gender.valueOf("NOT_KNOWN"));

    assertEquals(e.getGender(), Gender.NOT_KNOWN);
    assertEquals(e.getGender().getI18nKey(), Gender.NOT_KNOWN.getI18nKey());
    assertEquals(e.getGender().ordinal(), Gender.NOT_KNOWN.ordinal());
    assertEquals(e.getGender().getIsoCode(), Gender.NOT_KNOWN.getIsoCode());

    e.setGender(Gender.MALE);

    assertEquals(e.getGender(), Gender.MALE);
    assertEquals(e.getGender().getI18nKey(), Gender.MALE.getI18nKey());
    assertEquals(e.getGender().ordinal(), Gender.MALE.ordinal());
    assertEquals(e.getGender().getIsoCode(), Gender.MALE.getIsoCode());

    e.setGender(Gender.FEMALE);

    assertEquals(e.getGender(), Gender.FEMALE);
    assertEquals(e.getGender().getI18nKey(), Gender.FEMALE.getI18nKey());
    assertEquals(e.getGender().ordinal(), Gender.FEMALE.ordinal());
    assertEquals(e.getGender().getIsoCode(), Gender.FEMALE.getIsoCode());

    e.setGender(Gender.NOT_APPLICABLE);

    assertEquals(e.getGender(), Gender.NOT_APPLICABLE);
    assertEquals(e.getGender().getI18nKey(), Gender.NOT_APPLICABLE.getI18nKey());
    assertEquals(e.getGender().ordinal(), Gender.NOT_APPLICABLE.ordinal());
    assertEquals(e.getGender().getIsoCode(), Gender.NOT_APPLICABLE.getIsoCode());

    // test history
    employeeDao.update(e);
    List<DisplayHistoryEntry> historyEntriesAfter = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(historyEntriesBefore.size() + 1, historyEntriesAfter.size());

    final DisplayHistoryEntry genderHistoryEntry = historyEntriesAfter.get(0);
    assertEquals(genderHistoryEntry.getPropertyName(), "gender");
    assertEquals(genderHistoryEntry.getNewValue(), "[" + Gender.NOT_APPLICABLE.toString() + "]");
  }

  @Test
  public void testGenderConverter()
  {
    GenderConverter genderConverter = new GenderConverter();

    assertEquals(genderConverter.convertToDatabaseColumn(null).intValue(), Gender.NOT_KNOWN.getIsoCode());
    assertEquals(genderConverter.convertToDatabaseColumn(Gender.NOT_KNOWN).intValue(), Gender.NOT_KNOWN.getIsoCode());
    assertEquals(genderConverter.convertToDatabaseColumn(Gender.MALE).intValue(), Gender.MALE.getIsoCode());
    assertEquals(genderConverter.convertToDatabaseColumn(Gender.FEMALE).intValue(), Gender.FEMALE.getIsoCode());
    assertEquals(genderConverter.convertToDatabaseColumn(Gender.NOT_APPLICABLE).intValue(), Gender.NOT_APPLICABLE.getIsoCode());

    assertEquals(genderConverter.convertToEntityAttribute(null), Gender.NOT_KNOWN);
    assertEquals(genderConverter.convertToEntityAttribute(Integer.MAX_VALUE), Gender.NOT_KNOWN);
    assertEquals(genderConverter.convertToEntityAttribute(Integer.MIN_VALUE), Gender.NOT_KNOWN);
    assertEquals(genderConverter.convertToEntityAttribute(0), Gender.NOT_KNOWN);
    assertEquals(genderConverter.convertToEntityAttribute(1), Gender.MALE);
    assertEquals(genderConverter.convertToEntityAttribute(2), Gender.FEMALE);
    assertEquals(genderConverter.convertToEntityAttribute(9), Gender.NOT_APPLICABLE);
  }

  @Test
  public void testBanking()
  {
    log.info("Count employees: " + employeeList.size());
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    log.info("Employee: " + e.toString());
    List<DisplayHistoryEntry> historyEntriesBefore = employeeDao.getDisplayHistoryEntries(e);

    String iban = "/*/*/*/*/*///*/*//*//*/*/*/*/*/*/*/*/*/*/*/*/*////";
    e.setIban(iban);
    assertEquals(e.getIban(), iban);
    employeeDao.update(e); // for history test

    String bic = "ööäööäööäöä";
    e.setBic(bic);
    assertEquals(e.getBic(), bic);
    employeeDao.update(e); // for history test

    String accountHolder = "Mr. X";
    e.setAccountHolder(accountHolder);
    assertEquals(e.getAccountHolder(), accountHolder);
    employeeDao.update(e); // for history test

    // test history
    List<DisplayHistoryEntry> historyEntriesAfter = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(historyEntriesBefore.size() + 3, historyEntriesAfter.size());

    assertHistoryEntry(historyEntriesAfter.get(0), "accountHolder", accountHolder);
    assertHistoryEntry(historyEntriesAfter.get(1), "bic", bic);
    assertHistoryEntry(historyEntriesAfter.get(2), "iban", iban);
  }

  @Test
  public void testAddress()
  {
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    List<DisplayHistoryEntry> historyEntriesBefore = employeeDao.getDisplayHistoryEntries(e);

    String street = "Some street";
    e.setStreet(street);
    assertEquals(e.getStreet(), street);
    employeeDao.update(e);

    String zipCode = "12345";
    e.setZipCode(zipCode);
    assertEquals(e.getZipCode(), zipCode);
    employeeDao.update(e);

    String city = "Kassel";
    e.setCity(city);
    assertEquals(e.getCity(), city);
    employeeDao.update(e);

    String country = "Deutschland";
    e.setCountry(country);
    assertEquals(e.getCountry(), country);
    employeeDao.update(e);

    String state = "Hessen";
    e.setState(state);
    assertEquals(e.getState(), state);
    employeeDao.update(e);

    // test history
    List<DisplayHistoryEntry> historyEntriesAfter = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(historyEntriesBefore.size() + 5, historyEntriesAfter.size());

    assertHistoryEntry(historyEntriesAfter.get(0), "state", state);
    assertHistoryEntry(historyEntriesAfter.get(1), "country", country);
    assertHistoryEntry(historyEntriesAfter.get(2), "city", city);
    assertHistoryEntry(historyEntriesAfter.get(3), "zipCode", zipCode);
    assertHistoryEntry(historyEntriesAfter.get(4), "street", street);
  }

  @Test
  public void testStaffNumber()
  {
    assertTrue(employeeList.size() > 0);
    EmployeeDO e = employeeList.get(0);
    List<DisplayHistoryEntry> historyEntriesBefore = employeeDao.getDisplayHistoryEntries(e);

    String staffNumber = "123abc456def";
    e.setStaffNumber(staffNumber);
    assertEquals(e.getStaffNumber(), staffNumber);
    employeeDao.update(e);

    // test history
    List<DisplayHistoryEntry> historyEntriesAfter = employeeDao.getDisplayHistoryEntries(e);
    assertEquals(historyEntriesBefore.size() + 1, historyEntriesAfter.size());

    assertHistoryEntry(historyEntriesAfter.get(0), "staffNumber", staffNumber);
  }

  private void assertHistoryEntry(final DisplayHistoryEntry historyEntry, final String propertyName, final String newValue)
  {
    assertEquals(historyEntry.getPropertyName(), propertyName);
    assertEquals(historyEntry.getNewValue(), newValue);
  }

}

File: projectforge-business/src/test/kotlin/org/projectforge/business/employee/EmployeeTest.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2025 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.employee

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.projectforge.business.fibu.EmployeeDO
import org.projectforge.business.fibu.EmployeeDao
import org.projectforge.business.fibu.EmployeeService
import org.projectforge.business.user.GroupDao
import org.projectforge.business.user.ProjectForgeGroup
import org.projectforge.business.user.UserRightId
import org.projectforge.business.user.UserRightValue
import org.projectforge.framework.persistence.user.api.ThreadLocalUserContext
import org.projectforge.framework.persistence.user.entities.PFUserDO
import org.projectforge.framework.persistence.user.entities.UserRightDO
import org.projectforge.business.test.AbstractTestBase
import org.projectforge.business.test.HistoryTester
import org.springframework.beans.factory.annotation.Autowired
import java.math.BigDecimal
import java.time.LocalDate

class EmployeeTest : AbstractTestBase() {
    @Autowired
    private lateinit var employeeDao: EmployeeDao

    private lateinit var employeeList: List<EmployeeDO>

    @BeforeEach
    fun init() {
        logon(TEST_FULL_ACCESS_USER)
        employeeList = employeeDao.selectAll(checkAccess = false)
        Assertions.assertTrue(employeeList.isNotEmpty(), "Keine Mitarbeiter in der Test DB!")
    }

    @AfterEach
    fun clean() {
        //Get initial infos
        baseLog.info("Cleanup deleted employess -> undelete")
        baseLog.info("Count employees: " + employeeList.size)
        Assertions.assertTrue(employeeList.isNotEmpty())
        persistenceService.runInTransaction { _ ->
            for (e in employeeList) {
                baseLog.info("Employee: $e")
                if (e.deleted) {
                    //Undelete
                    employeeDao.undelete(e, checkAccess = false)
                }
            }
        }
    }

    @Test
    fun testMarkAsDeleted() {
        //Get initial infos
        baseLog.info("Count employees: " + employeeList.size)
        Assertions.assertTrue(employeeList.isNotEmpty())
        val e = employeeList[0]
        baseLog.info("Employee: $e")

        //Mark as deleted
        employeeDao.markAsDeleted(e)

        //Check updates
        val updatdEmployee = employeeDao.find(e.id)
        Assertions.assertTrue(updatdEmployee!!.deleted)
        employeeDao.update(e)

        // TODO: Test unique constraint violation
        // TODO: Insert deleted validity attr again (to be implemented)
    }

    @Test
    fun testStaffNumber() {
        Assertions.assertTrue(employeeList.isNotEmpty())
        val e = employeeList[0]
        val historyEntriesBefore = employeeDao.loadHistory(e).sortedEntries
        val staffNumber = "123abc456def"
        e.staffNumber = staffNumber
        Assertions.assertEquals(e.staffNumber, staffNumber)
        employeeDao.update(e)

        // test history
        val historyEntriesAfter = employeeDao.loadHistory(e).sortedEntries
        Assertions.assertEquals(historyEntriesBefore.size + 1, historyEntriesAfter.size)
        HistoryTester.assertHistoryAttr(historyEntriesAfter[0], "staffNumber", value = staffNumber, oldValue = null)

    }

    companion object {
        /**
         * @param email: Optional mail address of the user.
         */
        fun createEmployee(
            employeeService: EmployeeService, employeeDao: EmployeeDao, test: AbstractTestBase,
            name: String,
            hrAccess: Boolean = false,
            groupDao: GroupDao? = null,
            email: String? = null,
        ): EmployeeDO {
            val loggedInUser = ThreadLocalUserContext.loggedInUser
            test.logon(TEST_ADMIN_USER)
            val user = PFUserDO()
            val useName = "${test.javaClass.simpleName}.$name"
            user.firstname = "$name.firstname"
            user.lastname = "$name.lastname"
            user.username = "$useName.username"
            user.email = email
            if (hrAccess) {
                user.addRight(UserRightDO(UserRightId.HR_VACATION, UserRightValue.READWRITE))
            }
            test.initTestDB.addUser(user)
            if (hrAccess) {
                val group = test.getGroup(ProjectForgeGroup.HR_GROUP.toString())
                group.assignedUsers!!.add(user)
                groupDao!!.update(group)
            }
            val employee = EmployeeDO()
            employee.user = user
            employeeDao.insert(employee, checkAccess = false)
            employeeService.insertAnnualLeaveDays(employee, LocalDate.now().minusYears(2), BigDecimal(25), false)
            employeeService.insertAnnualLeaveDays(employee, LocalDate.now().minusYears(1), BigDecimal(30), false)
            loggedInUser?.let { test.logon(it) }
            return employee
        }
    }
}
